pbench_logparse/main.go

```diff
@@ -0,0 +1,319 @@
+var fileHeader = map[string][]string{}
+
+type stringSlice []string
+
+type host struct {
+	kind      string
+	resultDir string
+	results   []resultType
+}
+
+type resultType struct {
+	kind, path           string
+	min, max, avg, pct95 float64
+}
+
+func (s *stringSlice) String() string {
+	return fmt.Sprintf("%v", *s)
+}
+
+func (s *stringSlice) Set(value string) error {
+	*s = append(*s, value)
+	return nil
+}
+
+func initFlags() {
+	flag.StringVar(&searchDir, "i", "/var/lib/pbench-agent/benchmark_result/tools-default/", "pbench run result directory to parse")
+	flag.StringVar(&resultDir, "o", "/tmp/", "output directory for parsed CSV result data")
+	flag.StringVar(&processes, "proc", "openshift_start_master_api,openshift_start_master_controll,openshift_start_node,/etcd", "list of processes to gather")
+	flag.Var(&blockDevices, "blkdev", "List of block devices")
+	flag.Var(&netDevices, "netdev", "List of network devices")
+	flag.Parse()
+}
+
+func main() {
+	var hosts []host
+	// This regexp matches the prefix to each pbench host result directory name
+	// which indicates host type. (ie. svt-master-1:pbench-benchmark-001/)
+	hostRegex := regexp.MustCompile(`svt[_-][elmn]\w*[_-]\d`)
+	initFlags()
+
+	// Check if no flags were passed, print help
+	if flag.NFlag() == 0 {
+		flag.PrintDefaults()
+		return
+	}
+
+	// Ensure output directory has a trailing slash
+	if string(resultDir[len(resultDir)-1]) != "/" {
+		resultDir = resultDir + "/"
+	}
+
+	// If no block devices were passed, don't add to search
+	if len(blockDevices) > 0 {
+		fileHeader["disk_IOPS.csv"] = []string(blockDevices)
+	}
+
+	// If no network devices were passed, don't add to search
+	if len(netDevices) > 0 {
+		fileHeader["network_l2_network_packets_sec.csv"] = []string(netDevices)
+		fileHeader["network_l2_network_Mbits_sec.csv"] = []string(netDevices)
+	}
+
+	processList = strings.Split(processes, ",")
+	// If no process names were passed, don't add to search
+	if len(processList) > 0 {
+		fileHeader["cpu_usage_percent_cpu.csv"] = processList
+		fileHeader["memory_usage_resident_set_size.csv"] = processList
+	}
+
+	// Return director listing of searchDir
+	dirList, err := ioutil.ReadDir(searchDir)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Iterate over directory contents
+	for _, item := range dirList {
+		// Match subdirectory that follows our pattern
+		if hostRegex.MatchString(item.Name()) && item.IsDir() {
+			kind := strings.Split(item.Name(), ":")
+			newHost := host{
+				kind:      kind[0],
+				resultDir: searchDir + item.Name(),
+			}
+			hosts = append(hosts, newHost)
+		}
+	}
+
+	// Maps are not ordered, create ordered slice of keys
+	keys := []string{}
+	for k := range fileHeader {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	// Iterate over all known hosts
+	for i, host := range hosts {
+		fmt.Printf("Host: %+v\n", host)
+		// Find each raw data CSV
+		for _, key := range keys {
+			fmt.Println(key)
+			fileList := findFile(host.resultDir, key)
```

Comment:
Doesn't really matter? I guess the function is called `findFile` singular. Could just throw an error if the `len(fileList) > 1` if we really wanted?